File: Cpp/Solution_Day_2.cpp
#include<iostream>
using namespace std;

class node {
  public :
    int data;
    node *left, *right;
};

node *addNode(int k){
  node *temp = new node();
  
  temp -> data = k;
  temp -> left = NULL;
  temp -> right = NULL;
  
  return temp;
}

bool match(node *node, int key){
  if(node == NULL){
    return true;
  }
  
  return node -> data == key && match(node -> left, key) && match(node -> right, key);
}

bool checking(node *node){
  int key = node -> data;
  
  return match(node, key);
}

void check(node *node, int *count){
    if(node == NULL){
        return;
    }
    
    if(checking(node)){
        (*count)++;
    }
    
    check(node -> left, count);
    check(node -> right, count);
}

int start(node *node){
    int count = 0;
    
    check(node, &count);
    
    return count;
}

int main(){
  node *root = NULL;
  
  root = addNode(1);
  root -> left = addNode(1);
  root -> right = addNode(0);
  root -> left -> left = addNode(1);
  root -> left -> right = addNode(1);
  root -> right -> left = addNode(0);
  root -> right -> right = addNode(1);
  
  cout << start(root);
  return 0;
}

File: JS/Solution_Day_2.js
class node{
  constructor(key){
    this.data = key;
    this.left = null;
    this.right = null;
  }
}

let match = (root, key) => {
  if(root === null){
    return true;
  }
  
  return root.data === key && match(root.left, key) && match(root.right, key);
};

let matching = root => {
  let key = root.data;
  
  return match(root, key);
};

let checking = (root, branch) => {
  if(root === null){
    return;
  }
  
  if(matching(root)){
    branch.count++;
  }
  
  checking(root.left, branch);
  checking(root.right, branch);
};

let start = root => {
  let branch = { count : 0};
  count = checking(root, branch);
  return branch.count;
};

let head = null;

head = new node(1);
head.left = new node(1);
head.right = new node(0);
head.left.left = new node(1);
head.left.right = new node(1);
head.right.left = new node(0);
head.right.right = new node(1);

console.log(start(head));